src/main/java/emu/grasscutter/game/quest/GameMainQuest.java

```diff
@@ -125,9 +125,36 @@ public GameQuest getChildQuestByOrder(int order) {
     }
 
     public void finish() {
+        // Avoid recursion from child finish() in GameQuest
+        // when auto finishing all child quests with QUEST_STATE_UNFINISHED (below)
+        if (this.isFinished) {
+            Grasscutter.getLogger().debug("Skip main quest finishing because it's already finished");
+            return;
+        }
+
         this.isFinished = true;
         this.state = ParentQuestState.PARENT_QUEST_STATE_FINISHED;
 
+         /*
+            We also need to check for unfinished childQuests in this MainQuest
+            force them to complete and send a packet about this to the user,
+            because at some points there are special "invisible" child quests that control
+            some situations.
+
+            For example, subQuest 35312 is responsible for the event of leaving the territory
+            of the island with a statue and automatically returns the character back,
+            quest 35311 completes the main quest line 353 and starts 35501 from
+            new MainQuest 355 but if 35312 is not completed after the completion
+            of the main quest 353 - the character will not be able to leave place
+            (return again and again)
+            */
+        this
+            .getChildQuests()
+            .values()
+            .stream()
+            .filter(p -> p.state == QuestState.QUEST_STATE_UNFINISHED)
```

Comment:
It seems the game doesnt respond to packages about unstarted quests (I could be wrong), I will check and see if there will be bugs with packets with "unstarted" state

Reply:
Changed

Reply:
> woudln't it be better to check for != `QUEST_STATE_FINISHED`? Otherwise there might still be failed or unstarted quests left

official leaves failed and unstarted quests as is. It doesn't set them to finished

Reply:
> > woudln't it be better to check for != `QUEST_STATE_FINISHED`? Otherwise there might still be failed or unstarted quests left
> 
> official leaves failed and unstarted quests as is. It doesn't set them to finished

But what about quests that would normally trigger other hidden quests or similar things? 
Because we can skip quests by using commands, without completing them.